docs/src/index.md

```diff
@@ -6,4 +6,369 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+    At the end of the day they are all _wibbles_ or _wobbles_.
+
+### Self derivative `Δself`, `∂self` etc.
+
+!!! Terminology
+    To my knowledge there is no standard terminology for this.
+    Other good names might be `Δinternal`/`∂internal`
+
+From the mathematical perspective,
+one may have been wondering what all this `Δself`, `∂self` is.
+After all a function with two inputs:
+say `f(a, b)` only has two partial derivatives,
+``\dfrac{∂f}{∂a}``, ``\dfrac{∂f}{∂b}``,
+why then does the _pushforward_ take in this extra `Δself`,
+and why does the _pullback_ return this extra `∂self` ?
+
+The thing is in julia
+the function `f` may itself have internal values.
+For example a closure has the fields it closes over; and a callable object (i.e. a functor) like a `Flux.Dense` has the fields of that object.
+
+**Thus every function is treated as having the extra implicit argument `self`,
+which captures those fields.**
+So all _pushforward_ take in a extra argument,
+which unless they are for things with fields, they ignore. (thus common to write `function pushforward(_, Δargs...)` in those cases),
+and every _pullback_ return an extra `∂self`,
+which is, for things without fields, the constant `NO_FIELDS` which indicates there is no fields within the function itself.
+
+
+#### Pushforward / Pullback summary
+- **Pushforward:**
+    - returned by `ffrule`
+    - takes input space wibbles, gives output space wobbles
+    - 1 argument per orignal function argument + 1 for the function itself
+    - 1 return per orignal function return
+- **Pullback**
+   -  return by `rrule`
+   - takes output space wobbles, gives input space wibbles
+   - 1 argument per original function return
+   - 1 return per orignal function argument + 1 for the function itself
+
+#### Pushforward/Pullback and Total Derivative/Gradient
+
+The most trivial use of the frule+pushforward is to calculate the [Total Derivative](https://en.wikipedia.org/wiki/Total_derivative):
+```julia
+y, pushforward = frule(f, a, b, c)
+ẏ = pushforward(1, 1, 1, 1)  # for appropriate `1`-like perturbation.
+```
+Then we have that
+`ẏ` is the _total derivative_ of
+`f` at `(a, b, c)`:
+written mathematically as ``df_{(a,b,c)}``
+
+
+Similarly:
+The most trivial use of the rrule+pullback is to calculate the [Gradient](https://en.wikipedia.org/wiki/Gradient):
+```julia
+y, pullback = frule(f, a, b, c)
+∇f  = pushforward(1) # for appropriate `1`-like seed.
+s̄, ā, b̄, c̄ = ∇f
+```
+Then we have that
+`∇f` is the _gradient_ of
+`f` at `(a, b, c)`.
+And we thus have the partial derivative:
+s̄, ā, b̄, c̄.
+(Including the and the self-partial derivative,
+s̄).
+Written mathematically as ``\dfrac{∂f}{∂a}``, ``\dfrac{∂f}{∂b}``, ``\dfrac{∂f}{∂c}``.
+
+### Differentials
+
+The values that come back from pullbacks,
+or pushforwards
+are not always the same type as the input/outputs of the original function.
+They are differentials,
+differency-equivalents.
```

Comment:
Is this a standard term?

Reply:
nope.
I'm also not entirely convinced the way we use differentials is inline with standard terms.

this bit should be worked on